File: javascript/code-challenges/tree/classes/binary-tree.js
'use strict';

// local module
const Node = require('./node.js');

class BinaryTree {
  constructor() {
    this.root = null;
  }

  add(val) {
    let node = new Node(val);

    if (!this.root) {
      this.root = node;

      return console.log('Node value added: ', node.value);
    }

    let currentNode = this.root;
    let trackingNode = null;

    while (currentNode !== null) {
      trackingNode = currentNode;
      if (val < currentNode.value) {
        currentNode = currentNode.left;
      } else {
        currentNode = currentNode.right;
      }
    }

    if (val < trackingNode.value) {
      trackingNode.left = node;
    } else {
      trackingNode.right = node;
    }

    return console.log('Node value added: ', node.value);
  }

  preOrder() {
    if (!this.root) {
      return 'Empty Tree!';
    }

    let nodes = [];

    let _walk = (node) => {
      nodes.push(node.value);
      if (node.left) { _walk(node.left); }
      if (node.right) { _walk(node.right); }
    };

    _walk(this.root);

    return nodes;
  }

  inOrder() {
    if (!this.root) {
      return 'Empty Tree!';
    }

    let nodes = [];

    let _walk = (node) => {
      if (node.left) { _walk(node.left); }
      nodes.push(node.value);
      if (node.right) { _walk(node.right); }
    };

    _walk(this.root);

    return nodes;
  }

  postOrder() {
    if (!this.root) {
      return 'Empty Tree!';
    }

    let nodes = [];

    let _walk = (node) => {
      if (node.left) { _walk(node.left); }
      if (node.right) { _walk(node.right); }
      nodes.push(node.value);
    };

    _walk(this.root);

    return nodes;
  }

  breadthFirst() {
    if (!this.root) {
      return 'Empty Tree!';
    }

    let nodes = [];
    let queue = [];

    queue.push(this.root);

    while (queue.length > 0) {
      let currentNode = queue.shift();
      nodes.push(currentNode.value);

      if (currentNode.left) {
        queue.push(currentNode.left);
      }
      if (currentNode.right) {
        queue.push(currentNode.right);
      }
    }
    return nodes;
  }

  contains(value) {
    let currentNode = null;
    let containsNum = null;
    let _walk = (node) => {
      currentNode = node;

      if (currentNode === null) {
        containsNum = false;
        return false;
      }

      if (currentNode.value === value) {
        containsNum = true;
        return true;
      }

      let leftNodes = _walk(node.left);

      if (leftNodes === true) {
        containsNum = true;
        return true;
      }

      let rightNodes = _walk(node.right);

      if (rightNodes === true) {
        containsNum = true;
        return true;
      }
    };
    _walk(this.root);

    return containsNum;
  }

  findMaximumValue() {
    let max = this.root.value;

    let nodeChecker = (node) => {

      if (node.value > max) {
        max = node.value;
      }
      if (node.left) {
        nodeChecker(node.left);
      }
      if (node.right) {
        nodeChecker(node.right);
      }
    };
    nodeChecker(this.root);

    return max;
  }
}

module.exports = BinaryTree;

File: python/code_challenges/tree_intersection/binary_tree.py
from code_challenges.tree_intersection.tree_node import Node
from code_challenges.tree_intersection.queue import Queue

class BinaryTree():
    def __init__(self, root=None):
        if root != None:
            self.root = Node(root)
        else:
            self.root = None

    def __str__(self):
        return f'A Binary Tree, its root has a value of {self.root.value}'

    def __repr__(self):
        return f'BinaryTree(root: {self.root.value})'

    def pre_order(self):
        '''
        Returns a pre-ordered list of the tree's node values.

        Input: None
        Output: list
        '''
        if self.root == None:
            raise Exception('Empty Tree!')

        nodes = []
        def walk(node):
            nodes.append(node.value)
            if node.left: walk(node.left)
            if node.right: walk(node.right)

        walk(self.root)

        return nodes

    def in_order(self):
        '''
        Returns a in-ordered list of the tree's node values.

        Input: None
        Output: list
        '''
        if self.root == None:
            raise Exception('Empty Tree!')

        nodes = []
        def walk(node):
            if node.left: walk(node.left)
            nodes.append(node.value)
            if node.right: walk(node.right)

        walk(self.root)

        return nodes

    def post_order(self):
        '''
        Returns a post-ordered list of the tree's node values.

        Input: None
        Output: list
        '''
        if self.root == None:
            raise Exception('Empty Tree!')

        nodes = []
        def walk(node):
            if node.left: walk(node.left)
            if node.right: walk(node.right)
            nodes.append(node.value)

        walk(self.root)

        return nodes

    def max(self):
        '''
        Returns the largest value in the tree.

        Input: None
        Output: Number
        '''
        if self.root == None:
            raise Exception('Tree is empty!')

        max_value = self.root.value
        node_queue = Queue()
        node_queue.enqueue(self.root)

        while node_queue.is_empty() == False:
            if node_queue.front.value.left:
                node_queue.enqueue(node_queue.front.value.left)
            if node_queue.front.value.right:
                node_queue.enqueue(node_queue.front.value.right)

            temp_node = node_queue.dequeue()
            if temp_node.value > max_value:
                max_value = temp_node.value

        return max_value
